File: 980/980.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/980.py

fun buildMulTableQ8(): Array<IntArray> {
    val baseArr = Array(4) { IntArray(4) }
    val sgnArr = Array(4) { IntArray(4) { 1 } }
    for (a in 0 until 4) {
        for (b in 0 until 4) {
            if (a == 0) {
                baseArr[a][b] = b
                sgnArr[a][b] = 1
            } else if (b == 0) {
                baseArr[a][b] = a
                sgnArr[a][b] = 1
            } else if (a == b) {
                baseArr[a][b] = 0
                sgnArr[a][b] = -1
            } else {
                when {
                    a == 1 && b == 2 -> { baseArr[a][b] = 3; sgnArr[a][b] = 1 }
                    a == 2 && b == 3 -> { baseArr[a][b] = 1; sgnArr[a][b] = 1 }
                    a == 3 && b == 1 -> { baseArr[a][b] = 2; sgnArr[a][b] = 1 }
                    a == 2 && b == 1 -> { baseArr[a][b] = 3; sgnArr[a][b] = -1 }
                    a == 3 && b == 2 -> { baseArr[a][b] = 1; sgnArr[a][b] = -1 }
                    a == 1 && b == 3 -> { baseArr[a][b] = 2; sgnArr[a][b] = -1 }
                    else -> throw RuntimeException("Unexpected basis multiplication case")
                }
            }
        }
    }
    val mulTable = Array(8) { IntArray(8) }
    for (bigA in 0 until 8) {
        val sa = if (bigA < 4) 1 else -1
        val a = bigA and 3
        for (bigB in 0 until 8) {
            val sb = if (bigB < 4) 1 else -1
            val b = bigB and 3
            val s = sa * sb * sgnArr[a][b]
            val c = baseArr[a][b]
            mulTable[bigA][bigB] = if (s == 1) c else (c xor 4)
        }
    }
    return mulTable
}

fun buildRTable(mulTable: Array<IntArray>, genElems: IntArray): IntArray {
    val rTable = IntArray(8 * 3)
    for (v in 0 until 8) {
        for (b in 0 until 3) {
            rTable[v * 3 + b] = mulTable[v][genElems[b]]
        }
    }
    return rTable
}

fun buildInvTable(mulTable: Array<IntArray>): IntArray {
    val invTable = IntArray(8)
    for (e in 0 until 8) {
        for (f in 0 until 8) {
            if (mulTable[e][f] == 0 && mulTable[f][e] == 0) {
                invTable[e] = f
                break
            }
        }
    }
    return invTable
}

val MUL_TABLE = buildMulTableQ8()
val GEN_ELEMS = intArrayOf(1, 2, 7)
val R_TABLE = buildRTable(MUL_TABLE, GEN_ELEMS)
val INV_TABLE = buildInvTable(MUL_TABLE)

fun computeF(nVal: Int): Long {
    val MOD = 888888883L
    val MULT = 8888L
    var aVal = 88888888L
    val cnts = LongArray(8)
    val R = R_TABLE
    for (iter in 0 until nVal) {
        var v = 0
        for (step in 0 until 50) {
            v = R[v * 3 + (aVal % 3).toInt()]
            aVal = (aVal * MULT) % MOD
        }
        cnts[v]++
    }
    var total = 0L
    val invArr = INV_TABLE
    for (e in 0 until 8) {
        total += cnts[e] * cnts[invArr[e]]
    }
    return total
}

fun main() {
    val r10 = computeF(10)
    check(r10 == 13L) { "Assert F(10) == 13 failed, got $r10" }
    val r100 = computeF(100)
    check(r100 == 1224L) { "Assert F(100) == 1224 failed, got $r100" }
    val result = computeF(1_000_000)
    println(result)
}

File: 980/980.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/980.py

fn build_mul_table_q8() -> [[i32; 8]; 8] {
    let mut base_arr = [[0i32; 4]; 4];
    let mut sgn_arr = [[1i32; 4]; 4];
    for a in 0..4usize {
        for b in 0..4usize {
            if a == 0 {
                base_arr[a][b] = b as i32;
                sgn_arr[a][b] = 1;
            } else if b == 0 {
                base_arr[a][b] = a as i32;
                sgn_arr[a][b] = 1;
            } else if a == b {
                base_arr[a][b] = 0;
                sgn_arr[a][b] = -1;
            } else {
                match (a, b) {
                    (1, 2) => { base_arr[a][b] = 3; sgn_arr[a][b] = 1; }
                    (2, 3) => { base_arr[a][b] = 1; sgn_arr[a][b] = 1; }
                    (3, 1) => { base_arr[a][b] = 2; sgn_arr[a][b] = 1; }
                    (2, 1) => { base_arr[a][b] = 3; sgn_arr[a][b] = -1; }
                    (3, 2) => { base_arr[a][b] = 1; sgn_arr[a][b] = -1; }
                    (1, 3) => { base_arr[a][b] = 2; sgn_arr[a][b] = -1; }
                    _ => panic!("Unexpected basis multiplication case"),
                }
            }
        }
    }
    let mut mul_table = [[0i32; 8]; 8];
    for big_a in 0..8usize {
        let sa: i32 = if big_a < 4 { 1 } else { -1 };
        let a = big_a & 3;
        for big_b in 0..8usize {
            let sb: i32 = if big_b < 4 { 1 } else { -1 };
            let b = big_b & 3;
            let s = sa * sb * sgn_arr[a][b];
            let c = base_arr[a][b];
            mul_table[big_a][big_b] = if s == 1 { c } else { c ^ 4 };
        }
    }
    mul_table
}

fn build_r_table(mul_table: &[[i32; 8]; 8], gen_elems: &[usize; 3]) -> [i32; 24] {
    let mut r_table = [0i32; 24];
    for v in 0..8usize {
        for b in 0..3usize {
            r_table[v * 3 + b] = mul_table[v][gen_elems[b]];
        }
    }
    r_table
}

fn build_inv_table(mul_table: &[[i32; 8]; 8]) -> [usize; 8] {
    let mut inv_table = [0usize; 8];
    for e in 0..8usize {
        for f in 0..8usize {
            if mul_table[e][f] == 0 && mul_table[f][e] == 0 {
                inv_table[e] = f;
                break;
            }
        }
    }
    inv_table
}

fn compute_f(n_val: i64, r_table: &[i32; 24], inv_table: &[usize; 8]) -> i64 {
    const MOD: i64 = 888_888_883;
    const MULT: i64 = 8888;
    let mut a_val: i64 = 88_888_888;
    let mut cnts = [0i64; 8];
    for _iter in 0..n_val {
        let mut v: usize = 0;
        for _step in 0..50 {
            v = r_table[v * 3 + (a_val % 3) as usize] as usize;
            a_val = (a_val * MULT) % MOD;
        }
        cnts[v] += 1;
    }
    let mut total: i64 = 0;
    for e in 0..8usize {
        total += cnts[e] * cnts[inv_table[e]];
    }
    total
}

fn main() {
    let mul_table = build_mul_table_q8();
    let gen_elems: [usize; 3] = [1, 2, 7];
    let r_table = build_r_table(&mul_table, &gen_elems);
    let inv_table = build_inv_table(&mul_table);

    let r10 = compute_f(10, &r_table, &inv_table);
    assert_eq!(r10, 13, "Assert F(10) == 13 failed, got {}", r10);
    let r100 = compute_f(100, &r_table, &inv_table);
    assert_eq!(r100, 1224, "Assert F(100) == 1224 failed, got {}", r100);
    let result = compute_f(1_000_000, &r_table, &inv_table);
    println!("{}", result);
}
